src/proto/uprotocol/uuri.rs

```diff
@@ -11,38 +11,617 @@
+            }
+        }
+
+        let entity = UEntity {
+            name,
+            version_major: ve,
+            ..Default::default()
+        };
+
+        Ok(UUri {
+            entity: Some(entity).into(),
+            authority: authority.into(),
+            resource: resource.into(),
+            ..Default::default()
+        })
     }
 }
 
 impl TryFrom<&str> for UUri {
     type Error = SerializationError;
 
-    fn try_from(value: &str) -> Result<Self, Self::Error> {
-        LongUriSerializer::deserialize(value.into())
+    /// Attempts to serialize a `String` into a `UUri`.
+    ///
+    /// # Arguments
+    ///
+    /// * `uri` - The `String` to be converted into a `UUri`.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing either the `UUri` representation of the URI or a `SerializationError`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use up_rust::uprotocol::{UAuthority, UEntity, UResource};
+    /// use up_rust::uprotocol::UUri;
+    ///
+    /// let uri = UUri {
+    ///     entity: Some(UEntity {
+    ///         name: "example.com".to_string(),
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     resource: Some(UResource {
+    ///         name: "rpc".to_string(),
+    ///         instance: Some("raise".to_string()),
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     authority: None.into(),
+    ///     ..Default::default()
+    /// };
+    ///
+    /// let uri_from = UUri::try_from("/example.com//rpc.raise").unwrap();
+    /// assert_eq!(uri, uri_from);
+    /// ````
+    fn try_from(uri: &str) -> Result<Self, Self::Error> {
+        UUri::from_str(uri)
     }
 }
 
 impl TryFrom<UUri> for Vec<u8> {
     type Error = SerializationError;
 
     fn try_from(value: UUri) -> Result<Self, Self::Error> {
-        MicroUriSerializer::serialize(&value)
+        MicroUriSerializer::serialize(&value).map_err(|e| SerializationError::new(e.to_string()))
     }
 }
 
 impl TryFrom<Vec<u8>> for UUri {
     type Error = SerializationError;
 
     fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
-        MicroUriSerializer::deserialize(value)
+        MicroUriSerializer::deserialize(value).map_err(|e| SerializationError::new(e.to_string()))
+    }
+}
+
+impl UUri {
+    /// Builds a fully resolved `UUri` from the serialized long format and the serialized micro format.
+    ///
+    /// # Arguments
+    /// * `long_uri` - uri serialized as a string.
+    /// * `micro_uri` - uri serialized as a byte slice.
+    ///
+    /// # Returns
+    /// If successful, returns an UUri object serialized from the input formats. Returns `SerializationError` if the deserialization
+    /// fails or the resulting uri cannot be resolved.
+    pub fn build_resolved(long_uri: &str, micro_uri: &[u8]) -> Result<UUri, SerializationError> {
+        if long_uri.is_empty() && micro_uri.is_empty() {
```

Comment:
Maybe should be `||` instead with updated corresponding message noting one of them is empty _or_ alternatively (maybe better?) check each one independently so that the most descriptive error message can be returned, noting which one is empty.

Why? Well, I guess if either one of them is empty, we can't say that we're returning a fully resolved URI.

Reply:
Not arguing against your point - but should this behavior also be revised in [up-java](https://github.com/eclipse-uprotocol/up-java/blob/b502e03da95648c21e8f8b8daffe0ea082d69e82/src/main/java/org/eclipse/uprotocol/uri/serializer/UriSerializer.java#L66), then?

Reply:
Yeah, from my point of view `up-java` is not gospel, the spec is. Perhaps you could open an easy issue on `up-spec` confirming.

For example, from what I saw the more intensive checking on Micro URIs is not in `up-java`, but probably should be. I should go open an issue on that repo regarding this actually.